### PR TITLE
Fix 132.200 und 8.33 kHz Raster

### PR DESCRIPTION
**Pull request by tzobler**
Changes for rounding errors at frequencies 132.200 and 131.200 and correct handling of frequencies in the 8.33 kHz grid

### CODE REVIEW

File: src/plugins/simulator/fs9/simulatorfs9.cpp [hunk 1]
```diff
@@ -194,6 +194,15 @@ namespace BlackSimPlugin::Fs9
         return CSimulatorFsCommon::physicallyRemoveRemoteAircraft(callsign);
     }
 
+    bool CSimulatorFs9::dontOverwrite(const CFrequency &from, const CFrequency &to)
+    {
+        bool ergeb = false;
```
Comment (ltoenning) [inline src/plugins/simulator/fs9/simulatorfs9.cpp:199]: Avoid german abbreviations inside the code

File: src/plugins/simulator/fscommon/fsuipcimpl.cpp [hunk 2]
```diff
@@ -148,10 +148,10 @@ namespace BlackSimPlugin::FsCommon
...
         if (!this->isOpened()) { return false; }
 
-        quint16 com1ActiveRaw = static_cast<quint16>(aircraft.getCom1System().getFrequencyActive().value(CFrequencyUnit::MHz()) * 100);
-        quint16 com2ActiveRaw = static_cast<quint16>(aircraft.getCom2System().getFrequencyActive().value(CFrequencyUnit::MHz()) * 100);
-        quint16 com1StandbyRaw = static_cast<quint16>(aircraft.getCom1System().getFrequencyStandby().value(CFrequencyUnit::MHz()) * 100);
-        quint16 com2StandbyRaw = static_cast<quint16>(aircraft.getCom2System().getFrequencyStandby().value(CFrequencyUnit::MHz()) * 100);
+        quint16 com1ActiveRaw = static_cast<quint16>(aircraft.getCom1System().getFrequencyActive().value(CFrequencyUnit::kHz()) / 10);
+        quint16 com2ActiveRaw = static_cast<quint16>(aircraft.getCom2System().getFrequencyActive().value(CFrequencyUnit::kHz()) / 10);
+        quint16 com1StandbyRaw = static_cast<quint16>(aircraft.getCom1System().getFrequencyStandby().value(CFrequencyUnit::kHz()) / 10);
+        quint16 com2StandbyRaw = static_cast<quint16>(aircraft.getCom2System().getFrequencyStandby().value(CFrequencyUnit::kHz()) / 10);
```
Comment (ltoenning) [inline src/plugins/simulator/fscommon/fsuipcimpl.cpp:154]: The result of this conversion should be the same, right? What is the reason for this change?

File: src/plugins/simulator/fs9/simulatorfs9.h [hunk 3]
```diff
@@ -103,6 +103,8 @@ namespace BlackSimPlugin::Fs9
         QSharedPointer<CLobbyClient> m_lobbyClient;
 
         static constexpr int SkipUpdateCyclesForCockpit = 10;
+
+        bool dontOverwrite(const BlackMisc::PhysicalQuantities::CFrequency &from, const BlackMisc::PhysicalQuantities::CFrequency &to);
```
Comment (ltoenning) [inline src/plugins/simulator/fs9/simulatorfs9.h:107]: As this method is only used inside ``CSimulatorFs9::updateOwnAircraftFromSimulator``, we could continue using it as a local lambda.

File: src/plugins/simulator/fs9/simulatorfs9.cpp [hunk 4]
```diff
@@ -231,7 +240,23 @@ namespace BlackSimPlugin::Fs9
         if (newTransponder.getTransponderCode() != m_simTransponder.getTransponderCode()) { changed = true; }
         if (newTransponder.getTransponderMode() != m_simTransponder.getTransponderMode()) { changed = true; }
 
-        m_fsuipc->write(ownAircraft);
+        const CComSystem oldCom1 = getOwnComSystem(CComSystem::Com1);
+        const CComSystem oldCom2 = getOwnComSystem(CComSystem::Com2);
+
+        CSimulatedAircraft comAircraft;
+        comAircraft = ownAircraft;
```
Comment (ltoenning) [inline src/plugins/simulator/fs9/simulatorfs9.cpp:247]:
```suggestion
        CSimulatedAircraft comAircraft = ownAircraft;
```

File: src/plugins/simulator/fs9/simulatorfs9.cpp [hunk 5]
```diff
@@ -416,22 +441,23 @@ namespace BlackSimPlugin::Fs9
...
+            if (CSimulatorFs9::dontOverwrite(oldCom1.getFrequencyActive(), com1.getFrequencyActive()))
                 com1.setFrequencyActive(oldCom1.getFrequencyActive());
-            if (dontOverwrite(oldCom1.getFrequencyStandby(), com1.getFrequencyStandby()))
+            if (CSimulatorFs9::dontOverwrite(oldCom1.getFrequencyStandby(), com1.getFrequencyStandby()))
                 com1.setFrequencyStandby(oldCom1.getFrequencyStandby());
-            if (dontOverwrite(oldCom2.getFrequencyActive(), com2.getFrequencyActive()))
+            if (CSimulatorFs9::dontOverwrite(oldCom2.getFrequencyActive(), com2.getFrequencyActive()))
                 com2.setFrequencyActive(oldCom2.getFrequencyActive());
-            if (dontOverwrite(oldCom2.getFrequencyStandby(), com2.getFrequencyStandby()))
+            if (CSimulatorFs9::dontOverwrite(oldCom2.getFrequencyStandby(), com2.getFrequencyStandby()))
```
Comment (ltoenning) [inline src/plugins/simulator/fs9/simulatorfs9.cpp:460]: With this, can the user switch away from a 8.33 kHz frequency by only using the simulator? ``dontOverwrite`` only checks the first parameter if it is a 8.33 kHz frequency. Hence, if oldCom is a 8.33 kHz frequency tuned from within swift, all changes from the simulator will always be skipped.

File: src/plugins/simulator/fs9/simulatorfs9.cpp [hunk 6]
```diff
@@ -194,6 +194,15 @@ namespace BlackSimPlugin::Fs9
         return CSimulatorFsCommon::physicallyRemoveRemoteAircraft(callsign);
     }
 
+    bool CSimulatorFs9::dontOverwrite(const CFrequency &from, const CFrequency &to)
+    {
+        bool ergeb = false;
+        // don't overwrite 8.33 kHz frequencies
+        ergeb = (CComSystem::isExclusiveWithin8_33kHzChannel(from)); //&& from.valueInteger(CFrequencyUnit::kHz()) == 132200);
+        // && from.valueInteger(CFrequencyUnit::kHz()) / 10 == to.valueInteger(CFrequencyUnit::kHz()) / 10;
```
Comment (ltoenning) [inline src/plugins/simulator/fs9/simulatorfs9.cpp:202]: Is this code commented out intentionally?

File: src/plugins/simulator/fs9/simulatorfs9.cpp [hunk 7]
```diff
@@ -231,7 +240,23 @@ namespace BlackSimPlugin::Fs9
...
         if (newTransponder.getTransponderMode() != m_simTransponder.getTransponderMode()) { changed = true; }
 
-        m_fsuipc->write(ownAircraft);
+        const CComSystem oldCom1 = getOwnComSystem(CComSystem::Com1);
+        const CComSystem oldCom2 = getOwnComSystem(CComSystem::Com2);
+
+        CSimulatedAircraft comAircraft;
+        comAircraft = ownAircraft;
+
+        // if one of the frequencies in swift is in the 8.33 kHz ratser, do not overwrite com1 active/standby
```
Comment (ltoenning) [inline src/plugins/simulator/fs9/simulatorfs9.cpp:249]: Typo
```suggestion
        // if one of the frequencies in swift is in the 8.33 kHz raster, do not overwrite com1 active/standby
```

File: src/plugins/simulator/fs9/simulatorfs9.cpp [hunk 8]
```diff
@@ -231,7 +240,23 @@ namespace BlackSimPlugin::Fs9
...
+        if (!CComSystem::isExclusiveWithin8_33kHzChannel(newCom1.getFrequencyActive()) || !CComSystem::isExclusiveWithin8_33kHzChannel(newCom1.getFrequencyStandby()))
+        {
+            comAircraft.setCom1System(oldCom1);
+        }
+        // if one of the frequencies in swift is in the 8.33 kHz ratser, do not overwrite com2 active/standby
```
Comment (ltoenning) [inline src/plugins/simulator/fs9/simulatorfs9.cpp:254]: Typo
```suggestion
        // if one of the frequencies in swift is in the 8.33 kHz raster, do not overwrite com2 active/standby
```

**Review by ltoenning — CHANGES_REQUESTED** [review]
Is it required that we handled the frequency 132.200 differently with FS9? I have not found anything in the code about this except commented out code inside `` CSimulatorFs9::dontOverwrite``.

There are minor typos in the commit message (german and 3.88 -> 8.33). It would also be nice if the commit message mentions FS9 because this makes it easier to generate a changelog :smile:

File: src/plugins/simulator/fscommon/fsuipcimpl.cpp (hunk 2 above)
Comment (tzobler) [reply src/plugins/simulator/fscommon/fsuipcimpl.cpp:154]: The old version with MHz causes rounding errors at 131.200 and 132.200 and then gives 131.19 and 132.19 respectively

File: src/plugins/simulator/fs9/simulatorfs9.h (hunk 3 above)
Comment (tzobler) [reply src/plugins/simulator/fs9/simulatorfs9.h:107]: That's right, I planned to use this function elsewhere to check for 132,200 and 131,200. Change is revoked

File: src/plugins/simulator/fs9/simulatorfs9.cpp (hunk 1 above)
Comment (tzobler) [reply src/plugins/simulator/fs9/simulatorfs9.cpp:199]: this change can be removed again

File: src/plugins/simulator/fs9/simulatorfs9.cpp (hunk 4 above)
Comment (tzobler) [reply src/plugins/simulator/fs9/simulatorfs9.cpp:247]: will be changed

File: src/plugins/simulator/fs9/simulatorfs9.cpp (hunk 6 above)
Comment (tzobler) [reply src/plugins/simulator/fs9/simulatorfs9.cpp:202]: function completely removed

File: src/plugins/simulator/fs9/simulatorfs9.cpp (hunk 7 above)
Comment (tzobler) [reply src/plugins/simulator/fs9/simulatorfs9.cpp:249]: changed to "grid"

File: src/plugins/simulator/fs9/simulatorfs9.cpp (hunk 5 above)
Comment (tzobler) [reply src/plugins/simulator/fs9/simulatorfs9.cpp:460]: yes, that's exactly how it's intended. Only when 25kHz frequencies are set again in swift will the values be adopted by the simulator. This is part of the solution, otherwise the simulator will overwrite the 8.33 frequency with incorrect values.

Comment (ltoenning) [reply src/plugins/simulator/fs9/simulatorfs9.cpp:460]: Perhaps it would be a good idea to document the behavior of the non-8.33 kHz capable simulators in the user documentation. I will do this soon.

File: src/plugins/simulator/fs9/simulatorfs9.cpp [hunk 9]
```diff
@@ -418,9 +433,9 @@ namespace BlackSimPlugin::Fs9
         {
             const auto dontOverwrite = [](const CFrequency &from, const CFrequency &to) {
```
Comment (ltoenning) [inline src/plugins/simulator/fs9/simulatorfs9.cpp:434]: As the ``to`` parameter is unused now, it should be removed. This will also fix the compiler warning (unreferenced formal parameter).

**Review by ltoenning — COMMENTED** [review]
The commits of this PR should be squashed before merging (either by @tzobler or by the one who executes the merge).

File: src/plugins/simulator/fs9/simulatorfs9.cpp (hunk 9 above)
Comment (tzobler) [reply src/plugins/simulator/fs9/simulatorfs9.cpp]: done

**Review by ltoenning — APPROVED** [review]

**Comment by ltoenning** [comment]
@oktal3700 I pushed changes on top of this branch myself. Could you review them?

**Review by oktal3700 — APPROVED** [review]

**Review by ltoenning — APPROVED** [review]
